### PR TITLE
WIP: 1-D median filter implementation

### PR DESCRIPTION
**Pull request by syurkevi**
Adds a 1-d median filter addressing issue #1039 .
API still needs to be finalized

### CODE REVIEW

File: src/api/c/filters.cpp [hunk 1]
```diff
@@ -19,17 +20,68 @@
 using af::dim4;
 using namespace detail;
 
+af_err af_medfilt(af_array *out, const af_array in, const dim_t wind_length, const dim_t wind_width, const af_border_type edge_pad)
+{
+    return af_medfilt2(out, in, wind_length, wind_width, edge_pad);
```
Comment (pavanky) [inline src/api/c/filters.cpp]: Can you check the dimensions of `in` like @shehzan10 mentioned and call `af_medfilt1` if it is 1D ?

File: src/api/cpp/filters.cpp [hunk 2]
```diff
@@ -16,8 +17,20 @@ namespace af
 
 array medfilt(const array& in, const dim_t wind_length, const dim_t wind_width, const borderType edge_pad)
 {
+    return medfilt2(in, wind_length, wind_width, edge_pad);
```
Comment (pavanky) [inline src/api/cpp/filters.cpp]: call `af_medfilt` like it did before. The C API will punt to the proper function.

File: src/backend/cpu/kernel/medfilt.hpp [hunk 3]
```diff
@@ -18,7 +18,89 @@ namespace kernel
...
+                                    break;
+                                case AF_PAD_SYM:
+                                    wind_vals.push_back(in_ptr[im_roff]);
+                                    break;
+                            }
+                        } else {
+                            wind_vals.push_back(in_ptr[im_roff]);
+                        }
+                    }
+
```
Comment (pavanky) [inline src/backend/cpu/kernel/medfilt.hpp]: Do you really need two switch cases here? Can they perhaps be merged?

File: src/backend/cuda/kernel/medfilt.hpp [hunk 4]
```diff
@@ -192,13 +322,35 @@ void medfilt(Param<T> out, CParam<T> in, int w_len, int w_wid)
...
+    dim3 blocks(blk_x*in.dims[1], in.dims[2], in.dims[3] );
+
+    switch(w_wid) {
+        case  3: CUDA_LAUNCH((medfilt1<T,pad, 3>), blocks, threads, out, in, blk_x); break;
+        case  5: CUDA_LAUNCH((medfilt1<T,pad, 5>), blocks, threads, out, in, blk_x); break;
+        case  7: CUDA_LAUNCH((medfilt1<T,pad, 7>), blocks, threads, out, in, blk_x); break;
+        case  9: CUDA_LAUNCH((medfilt1<T,pad, 9>), blocks, threads, out, in, blk_x); break;
+        case 11: CUDA_LAUNCH((medfilt1<T,pad,11>), blocks, threads, out, in, blk_x); break;
+        case 13: CUDA_LAUNCH((medfilt1<T,pad,13>), blocks, threads, out, in, blk_x); break;
+        case 15: CUDA_LAUNCH((medfilt1<T,pad,15>), blocks, threads, out, in, blk_x); break;
```
Comment (pavanky) [inline src/backend/cuda/kernel/medfilt.hpp]: Does this really need to be templated?

File: src/backend/cuda/kernel/medfilt.hpp [hunk 5]
```diff
@@ -192,13 +322,35 @@ void medfilt(Param<T> out, CParam<T> in, int w_len, int w_wid)
...
-        case 11: CUDA_LAUNCH((medfilt<T,pad,11,11>), blocks, threads, out, in, blk_x, blk_y); break;
-        case 13: CUDA_LAUNCH((medfilt<T,pad,13,13>), blocks, threads, out, in, blk_x, blk_y); break;
-        case 15: CUDA_LAUNCH((medfilt<T,pad,15,15>), blocks, threads, out, in, blk_x, blk_y); break;
+        case  3: CUDA_LAUNCH((medfilt2<T,pad, 3, 3>), blocks, threads, out, in, blk_x, blk_y); break;
+        case  5: CUDA_LAUNCH((medfilt2<T,pad, 5, 5>), blocks, threads, out, in, blk_x, blk_y); break;
+        case  7: CUDA_LAUNCH((medfilt2<T,pad, 7, 7>), blocks, threads, out, in, blk_x, blk_y); break;
+        case  9: CUDA_LAUNCH((medfilt2<T,pad, 9, 9>), blocks, threads, out, in, blk_x, blk_y); break;
+        case 11: CUDA_LAUNCH((medfilt2<T,pad,11,11>), blocks, threads, out, in, blk_x, blk_y); break;
+        case 13: CUDA_LAUNCH((medfilt2<T,pad,13,13>), blocks, threads, out, in, blk_x, blk_y); break;
+        case 15: CUDA_LAUNCH((medfilt2<T,pad,15,15>), blocks, threads, out, in, blk_x, blk_y); break;
```
Comment (pavanky) [inline src/backend/cuda/kernel/medfilt.hpp]: I just realized that this is broken. An assumption is being made that `w_len == w_wid`. Can you remove the templates and pass in the values as is and check the peformance drop for `3x3` and `5x5` for various sizes?

File: src/backend/cuda/kernel/medfilt.hpp (hunk 4 above)
Comment (pavanky) [inline src/backend/cuda/kernel/medfilt.hpp]: 1D stuff median filter need not be limited to 15. Make it bigger (and remove templates)

File: src/backend/opencl/medfilt.cpp [hunk 6]
```diff
@@ -19,7 +19,29 @@ namespace opencl
...
+        case  7: kernel::medfilt1<T, pad,  7>(out, in); break;
+        case  9: kernel::medfilt1<T, pad,  9>(out, in); break;
+        case 11: kernel::medfilt1<T, pad, 11>(out, in); break;
+        case 13: kernel::medfilt1<T, pad, 13>(out, in); break;
+        case 15: kernel::medfilt1<T, pad, 15>(out, in); break;
+    }
+
+    return out;
+}
+
```
Comment (pavanky) [inline src/backend/opencl/medfilt.cpp]: Same issues as CUDA backend.

File: src/backend/opencl/medfilt.cpp [hunk 7]
```diff
@@ -28,20 +50,22 @@ Array<T> medfilt(const Array<T> &in, dim_t w_len, dim_t w_wid)
...
 }
 
-#define INSTANTIATE(T)\
-    template Array<T> medfilt<T, AF_PAD_ZERO     >(const Array<T> &in, dim_t w_len, dim_t w_wid); \
-    template Array<T> medfilt<T, AF_PAD_SYM>(const Array<T> &in, dim_t w_len, dim_t w_wid);
+#define INSTANTIATE(T)                                                                          \
+    template Array<T> medfilt1<T, AF_PAD_ZERO>(const Array<T> &in, dim_t w_wid);              \
+    template Array<T> medfilt1<T, AF_PAD_SYM >(const Array<T> &in, dim_t w_wid);              \
+    template Array<T> medfilt2<T, AF_PAD_ZERO>(const Array<T> &in, dim_t w_len, dim_t w_wid);    \
+    template Array<T> medfilt2<T, AF_PAD_SYM>(const Array<T> &in, dim_t w_len, dim_t w_wid);
```
Comment (pavanky) [inline src/backend/opencl/medfilt.cpp]: Same issues as CUDA backend.

File: src/backend/cuda/kernel/medfilt.hpp (hunk 5 above)
Comment (pavanky) [inline src/backend/cuda/kernel/medfilt.hpp]: If that's not possible, atleast error out upstream for non square cases.

**Comment by pavanky** [comment]
build arrayfire linux ci
build arrayfire osx ci
build arrayfire windows ci

**Comment by shehzan10** [comment]
build arrayfire linux ci
build arrayfire osx ci
build arrayfire windows ci

**Comment by shehzan10** [comment]
build arrayfire linux ci
build arrayfire osx ci
build arrayfire windows ci

**Pull request merged**